internal/libs/autofile/autofile.go

```diff
@@ -99,11 +100,22 @@ func (af *AutoFile) Close() error {
 	return af.closeFile()
 }
 
+func (af *AutoFile) needsFlushing() bool {
+	af.mtx.Lock()
+	defer af.mtx.Unlock()
+
+	return af.hasWrites
+}
+
 func (af *AutoFile) closeFileRoutine() {
 	for {
 		select {
 		case <-af.closeTicker.C:
-			_ = af.closeFile()
+			if af.needsFlushing() {
```

Comment:
seems like we're letting go of the lock in needs flushing and then acquiring it again which seems like a logical race if not an actual race.